### PR TITLE
feat: gw7697 linking growi user with slack

### PR DESCRIPTION
**Pull request by mudana-grune**
https://youtrack.weseek.co.jp/issue/GW-7697

### CODE REVIEW

File: packages/app/src/server/models/user.js [hunk 1]
```diff
@@ -689,6 +690,13 @@ module.exports = function(crowi) {
     user.save();
   };
 
+  userSchema.statics.findUserBySlackId = function(slackId) {
+    if (slackId == null) {
+      return Promise.resolve(null);
+    }
+    return this.findOne({ slackId });
+  };
+
```
Comment (hakumizuki) [inline packages/app/src/server/models/user.js:699]: Please use `async` function instead.

Comment (hakumizuki) [reply packages/app/src/server/models/user.js:699]: Additionally, I think there is no use case that we call `findUserBySlackId` for now due to the query optimization issue(we can reduce the number of query by fetching all corresponding growi users at once). So, please implement `findUsersBySlackIds` method that takes an array of slack ID as a parameter and returns an array of growi users.

File: packages/app/src/server/service/slack-command-handler/keep.js [hunk 2]
```diff
@@ -137,10 +139,27 @@ module.exports = (crowi) => {
...
+    await Promise.all(messages.map(async(message) => {
+      const growiUser = await User.findUserBySlackId(message.user);
+      if (growiUser) {
+        message.user = `${growiUser.name} (@${growiUser.username})`;
+      }
+      else {
+        message.user = `This slack ID is not registered (${message.user})`;
+      }
+    }));
+  };
```
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/keep.js:156]: This part has a query optimization issue(we can reduce the number of query by fetching all corresponding growi users at once). So, please implement `findUsersBySlackIds` method that takes an array of slack ID as a parameter and returns an array of growi users, and call it here.

The logic of `if-else` statement looks great.

File: packages/app/src/server/service/slack-command-handler/keep.js [hunk 3]
```diff
@@ -32,8 +33,9 @@ module.exports = (crowi) => {
   handler.createPage = async function(client, payload, interactionPayloadAccessor, respondUtil) {
     let result = [];
     const channelId = payload.channel.id; // this must exist since the type is always block_actions
-    const userChannelId = payload.user.id;
+    const user = await User.findUserBySlackId(payload.user.id);
 
+    const userChannelId = user ? user._id : null;
```
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/keep.js:38]: Thanks for improving this part as well. This looks great.

File: packages/app/src/server/service/slack-command-handler/create-page-service.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ class CreatePageService {
     const normalizedPath = pathUtils.normalizePath(sanitizedPath);
 
     // generate a dummy id because Operation to create a page needs ObjectId
-    const dummyObjectIdOfUser = new mongoose.Types.ObjectId();
+    const dummyObjectIdOfUser = userChannelId || new mongoose.Types.ObjectId();
```
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/create-page-service.js:23]: Why is this change necessary?

Comment (mudana-grune) [reply packages/app/src/server/service/slack-command-handler/create-page-service.js:23]: This is so that the page creator can also be converted based on slackID, if there is no slackID then the page creator will be 'Unknown'

Comment (hakumizuki) [reply packages/app/src/server/service/slack-command-handler/create-page-service.js:23]: OK. But the type of `dummyObjectIdOfUser` must be type of User Document which will be like `{ _id: ObjectId }` at least.

Comment (hakumizuki) [reply packages/app/src/server/service/slack-command-handler/create-page-service.js:23]: OK, so I think the variable name should be `userId`. Also, I want you to always use null check to compare if the object is truthy or not. So, the code should be like `userId != null ? userId : new mongoose.Types.ObjectId();`

File: packages/app/src/server/service/slack-command-handler/keep.js [hunk 5]
```diff
@@ -165,7 +198,7 @@ module.exports = (crowi) => {
   };
 
   handler.keepCreatePageAndSendPreview = async function(client, interactionPayloadAccessor, path, userChannelId, contentsBody, respondUtil) {
-    await createPageService.createPageInGrowi(interactionPayloadAccessor, path, contentsBody, respondUtil);
+    await createPageService.createPageInGrowi(interactionPayloadAccessor, path, contentsBody, respondUtil, userChannelId);
```
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/keep.js:201]: Please rename here as well.

File: packages/app/src/server/service/slack-command-handler/keep.js (hunk 3 above)
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/keep.js:38]: Please rename here as well

File: packages/app/src/server/service/slack-command-handler/create-page-service.js [hunk 6]
```diff
@@ -11,7 +11,7 @@ class CreatePageService {
     this.crowi = crowi;
   }
 
-  async createPageInGrowi(interactionPayloadAccessor, path, contentsBody, respondUtil) {
+  async createPageInGrowi(interactionPayloadAccessor, path, contentsBody, respondUtil, userChannelId) {
```
Comment (hakumizuki) [inline packages/app/src/server/service/slack-command-handler/create-page-service.js:14]: Please rename here as well

File: packages/app/src/server/service/slack-command-handler/keep.js [hunk 7]
```diff
@@ -32,8 +33,9 @@ module.exports = (crowi) => {
   handler.createPage = async function(client, payload, interactionPayloadAccessor, respondUtil) {
     let result = [];
     const channelId = payload.channel.id; // this must exist since the type is always block_actions
-    const userChannelId = payload.user.id;
+    const user = await User.findUserBySlackId(payload.user.id);
 
+    const userId = user ? user._id : null;
```
[the remark this answers is not in the document]
Comment (mudana-grune) [reply packages/app/src/server/service/slack-command-handler/keep.js]: just updated in
https://github.com/weseek/growi/pull/5300/commits/4d85d8aae1457e9176a6c9bd0ff0971c4aadd376
